pkg/controller/complianceremediation/complianceremediation_controller.go

```diff
@@ -134,32 +134,105 @@ func (r *ReconcileComplianceRemediation) reconcileRemediation(instance *compv1al
+	objectLogger := logger.WithValues("Object.Name", obj.GetName(), "Object.Namespace", obj.GetNamespace())
+
+	objectLogger.Info("Reconciling remediation object")
+
 	found := obj.DeepCopy()
 	err := r.client.Get(context.TODO(), types.NamespacedName{Name: obj.GetName(), Namespace: obj.GetNamespace()}, found)
-	if instance.Spec.Apply {
-		logger.Info("Remediation will be applied", "ComplianceRemediation.Name", instance.Name)
-		if errors.IsNotFound(err) {
-			return r.client.Create(context.TODO(), obj)
-		} else if meta.IsNoMatchError(err) {
-			// If the kind is not available in the cluster, we can't retry
-			return common.NewNonRetriableCtrlError("Unable to create ComplianceRemediation. Make sure the CRD is installed: %s", err)
+
+	if errors.IsForbidden(err) {
+		return common.NewNonRetriableCtrlError(
+			"Unable to get fix object from ComplianceRemediation. "+
+				"Please update the compliance-operator's permissions: %s", err)
+	} else if meta.IsNoMatchError(err) {
+		// If the kind is not available in the cluster, we can't retry
+		return common.NewNonRetriableCtrlError(
+			"Unable to get fix object for ComplianceRemediation. "+
+				"Make sure the CRD is installed: %s", err)
+	} else if errors.IsNotFound(err) {
+		if instance.Spec.Apply {
+			return r.doCreateGenericRemediation(obj, found, objectLogger)
 		}
-		// TODO(jaosorior): If the object is found, should we update it?
+
+		objectLogger.Info("The object wasn't found, so no action is needed to unapply it")
+		return nil
+	} else if err != nil {
 		return err
-	} else {
-		logger.Info("Remediation will be unapplied", "ComplianceRemediation.Name", instance.Name)
 	}
 
-	// unapply remediation
-	if err == nil {
-		return r.client.Delete(context.TODO(), obj)
-	} else if errors.IsNotFound(err) {
+	if instance.Spec.Apply {
+		return r.doPatchGenericRemediation(obj, found, objectLogger)
+	}
+
+	return r.doDeleteGenericRemediation(obj, found, objectLogger)
+}
+
+func (r *ReconcileComplianceRemediation) doCreateGenericRemediation(remObj *unstructured.Unstructured, foundObj *unstructured.Unstructured, logger logr.Logger) error {
+	logger.Info("Remediation will be created")
+	compv1alpha1.AddRemediationAnnotation(remObj)
+
+	createErr := r.client.Create(context.TODO(), remObj)
+
+	if errors.IsForbidden(createErr) {
+		// If the kind is not available in the cluster, we can't retry
+		return common.NewNonRetriableCtrlError(
+			"Unable to create fix object from ComplianceRemediation. "+
+				" Please update the compliance-operator's permissions: %s", createErr)
+	} else if errors.IsAlreadyExists(createErr) {
 		return nil
-	} else if meta.IsNoMatchError(err) {
+	}
+
+	return createErr
+}
+
+func (r *ReconcileComplianceRemediation) doPatchGenericRemediation(remObj *unstructured.Unstructured, foundObj *unstructured.Unstructured, logger logr.Logger) error {
+	logger.Info("Remediation patch object")
+	// ensure labels are set so we don't overwrite them
+	labels := foundObj.GetLabels()
```

Comment:
I'm sorry, I don't understand this branch. Why are we setting the labels on the found object to an empty map? It doesn't seem like we're using the found object further on in this function or in the caller where found is scoped? Shouldn't the merge strategy take care about merging the labels and annotations as well?

Reply:
Either way it might be nice to add a test about updating the labels, that would make the case clearer.

Reply:
I was having issues because the labels and annotations were unset with the PATCH... but maybe it's not an issue now since I switched to the Merge strategy.